Format revamped Tutorial 1 to highlight the CLI vs JSON approaches (#1039)
ivadomed/ivadomed: docs/source/tutorials/one_class_segmentation_2d_unet.rst

                            "scale": [0.1, 0.1],
                            "translate": [0.03, 0.03],
                            "applied_to": ["im", "gt"],
                            "dataset_type": ["training"]
                        },
                        "ElasticTransform": {
                            "alpha_range": [28.0, 30.0],
                            "sigma_range":  [3.5, 4.5],
                            "p": 0.1,
                            "applied_to": ["im", "gt"],
                            "dataset_type": ["training"]
                        },
                      "NormalizeInstance": {"applied_to": ["im"]}
                    }
                }

    From this point onward, we will discuss some of the key parameters to perform a one-class 2D
    segmentation training. Most parameters are configurable only via modification of the configuration ``JSON file``.
    For those that supports command line run time configuration, we included the respective command versions under the ``Command Line Interface`` tab

    - ``command``: Action to perform. Here, we want to **train** a model:

        .. tabs::

            .. tab:: JSON File

>                 We can set the field within the newly copied ``config.json`` file as follow, around `this line: <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L2>`__:

Reviewer: ```suggestion
                We can set the field within the newly copied ``config.json`` file as follow, around `this line <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L2>`__:
```

                .. code-block:: json

                    "command": "train"

            .. tab:: Command Line Interface

                Note that you can also pass this argument via CLI (see `Usage <../usage.html>`__)

                .. code-block:: bash

                        ivadomed --train -c path/to/config

    - ``path_output``: Folder name that will contain the output files (e.g., trained model, predictions, results).

        .. tabs::

            .. tab:: JSON File

>                 Usually around `this line: <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L4>`__ in the ``config.json`` is where you can update the ``path_output``.

Reviewer: ```suggestion
                Usually around `this line <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L4>`__ in the ``config.json`` is where you can update the ``path_output``.
```

                .. code-block:: json

                    "path_output": "spineGeneric"

            .. tab:: Command Line Interface

                Note that you can also pass this argument via CLI (see `Usage <../usage.html>`__)

                .. code-block:: bash

                    ivadomed -c path/to/config --path-output path/to/output/directory

    - ``loader_parameters:path_data``: Location of the dataset. As discussed in `Data <../data.html>`__, the dataset
      should conform to the BIDS standard. Modify the path so it points to the location of the downloaded dataset.

        .. tabs::

            .. tab:: JSON File

>                 Usually around **Line Number 12** in the ``config.json`` is where you can update the ``path_data`` within the ``loader_parameters`` sub-dictionary.

Reviewer: There should be a link to the file instead of a specific line.

Author: Good eye! I think missed this link somehow.

                .. code-block:: json

                    "path_data": "data_example_spinegeneric"

            .. tab:: Command Line Interface

                Note that you can also pass this argument via CLI (see `Usage <../usage.html>`__)

                  .. code-block:: bash

                    ivadomed -c path/to/config --path-data path/to/bids/data

    - ``loader_parameters:target_suffix``: Suffix of the ground truth segmentation. The ground truth is located
      under the ``DATASET/derivatives/labels`` folder. In our case, the suffix is ``_seg-manual``:

        .. tabs::

            .. tab:: JSON File

>                 Usually around `this line: <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L14>`__ in the ``config.json`` is where you can update the ``target_suffix`` within the ``loader_parameters`` sub-dictionary.

Reviewer: ```suggestion
                Usually around `this line <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L14>`__ in the ``config.json`` is where you can update the ``target_suffix`` within the ``loader_parameters`` sub-dictionary.
```

                .. code-block:: json

                    "target_suffix": ["_seg-manual"]

    - ``loader_parameters:contrast_params``: Contrast(s) of interest

        .. tabs::

            .. tab:: JSON File

>                 Usually `this line: <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L20>`__ in the ``config.json`` is where you can update the ``contrast_params`` sub-dictionary within the ``loader_parameters`` sub-dictionary.

Reviewer: ```suggestion
                Usually `this line <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L20>`__ in the ``config.json`` is where you can update the ``contrast_params`` sub-dictionary within the ``loader_parameters`` sub-dictionary.
```

                .. code-block:: json

                    "contrast_params": {
                         "training_validation": ["T1w", "T2w", "T2star"],
                         "testing": ["T1w", "T2w", "T2star"],
                         "balance": {}
                    }

    - ``loader_parameters:slice_axis``: Orientation of the 2D slice to use with the model.

        .. tabs::

            .. tab:: JSON File

>                 Usually around `this line: <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L29>`__ in the ``config.json`` is where you can update the ``slice_axis`` subkey within the ``loader_parameters`` sub-dictionary.

Reviewer: ```suggestion
                Usually around `this line <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L29>`__ in the ``config.json`` is where you can update the ``slice_axis`` subkey within the ``loader_parameters`` sub-dictionary.
```

                .. code-block:: json

                    "slice_axis": "axial"

    - ``loader_parameters:multichannel``: Turn on/off multi-channel training. If ``true``, each sample has several
      channels, where each channel is an image contrast. If ``false``, only one image contrast is used per sample.

        .. tabs::

            .. tab:: JSON File

>                 Usually around `this line: <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L30>`__ in the ``config.json`` is where you can update the ``multichannel`` subkey within the ``loader_parameters`` sub-dictionary.

Reviewer: ```suggestion
                Usually around `this line <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L30>`__ in the ``config.json`` is where you can update the ``multichannel`` subkey within the ``loader_parameters`` sub-dictionary.
```

                .. code-block:: json

                    "multichannel": false

                .. note::

                    The multichannel approach requires that for each subject, the image contrasts are co-registered. This implies that
                    a ground truth segmentation is aligned with all contrasts, for a given subject. In this tutorial, only one channel
                    will be used.

    - ``training_time:num_epochs``: the maximum number of epochs that will be run during training. Each epoch is composed
      of a training part and an evaluation part. It should be a strictly positive integer.

        .. tabs::

            .. tab:: JSON File

>                 Usually around `this line: <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L29>`__ in the ``config.json`` is where you can update the ``slice_axis`` subkey within the ``loader_parameters`` sub-dictionary.

Reviewer: ```suggestion
                Usually around `this line <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L29>`__ in the ``config.json`` is where you can update the ``slice_axis`` subkey within the ``loader_parameters`` sub-dictionary.
```

                .. code-block:: json

                    "num_epochs": 100

Train model
-----------

    Once the configuration file is ready, run the training:

    .. tabs::

        .. tab:: Command Line Interface

            .. code-block:: bash

               ivadomed --train -c config.json --path-data path/to/bids/data --path-output path/to/output/directory

>     - In the above command, we execute the ``train`` command and manually specified ``path-data`` and ``path-output`` and overwrote/replace the specification in ``config.json``

Reviewer: ```suggestion
    - In the above command, we execute the ``--train`` command and manually specified ``--path-data`` and ``--path-output`` and overwrote/replace the specification in ``config.json``
```
The "--" would help to distinguish between the command and the config field (ex: path-data with a dash and path_data with an underscore).

Author: Good point. I concur.

                            "wspace": 0.75,
                            "dspace": 1
                        },
                        "CenterCrop": {
                            "size": [128, 128]},
                        "RandomAffine": {
                            "degrees": 5,
                            "scale": [0.1, 0.1],
                            "translate": [0.03, 0.03],
                            "applied_to": ["im", "gt"],
                            "dataset_type": ["training"]
                        },
                        "ElasticTransform": {
                            "alpha_range": [28.0, 30.0],
                            "sigma_range":  [3.5, 4.5],
                            "p": 0.1,
                            "applied_to": ["im", "gt"],
                            "dataset_type": ["training"]
                        },
                      "NormalizeInstance": {"applied_to": ["im"]}
                    }
                }

    From this point onward, we will discuss some of the key parameters to perform a one-class 2D
    segmentation training. Most parameters are configurable only via modification of the configuration ``JSON file``.

>     sFor those that supports commandline run time configuration, we included the respective command versions under the ``Command Line Interface`` tab

Reviewer: ```suggestion
    For those that supports command line run time configuration, we included the respective command versions under the ``Command Line Interface`` tab
```

Author: Thanks! Too many Ctrl+ S. My bad. 👍 Fixed.

                .. code-block:: json

                    "num_epochs": 100

Train model
-----------

    Once the configuration file is ready, run the training:

    .. tabs::

        .. tab:: Command Line Interface

            .. code-block:: bash

               ivadomed --train -c config.json --path-data path/to/bids/data --path-output path/to/output/directory

    - We can pass other flags to execute different commands (training, testing, segmentation), see `Usage <../usage.html>`__.

    - ``--path-output``: Folder name that will contain the output files (e.g., trained model, predictions, results).

        .. tabs::

            .. tab:: Command Line Interface

                  .. code-block:: bash

                     --path-output path/to/output/directory

    - ``--path-data``: Location of the dataset. As discussed in `Data <../data.html>`__, the dataset
      should conform to the BIDS standard. Modify the path so it points to the location of the downloaded dataset.

        .. tabs::

            .. tab:: Command Line Interface

                .. code-block:: bash

>                     --path-data path/to/bids/data

Reviewer: This part (lines 351-373) is duplicated information from the block above (lines 345-349).

I think we want to have only two options here:
- CLI with  `--train`, `--path-data`, and `--path-output` flags
- CLI without the flags

Author: Great point, I didn't audit the actually content thoroughly enough, this is definitely more confusion than it needs to be. 

Streamlined further to this: 
![image](https://user-images.githubusercontent.com/6720068/149809635-6d2d8ce2-5de3-46e4-a0b8-de62c46ad98c.png)

                Note that you can also pass this argument via CLI (see `Usage <../usage.html>`__)

                .. code-block:: bash

                        ivadomed --train -c path/to/config

    - ``path_output``: Folder name that will contain the output files (e.g., trained model, predictions, results).

        .. tabs::

            .. tab:: JSON File

                At `this line <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L4>`__ in the ``config.json`` is where you can update the ``path_output``.

                .. code-block:: json

                    "path_output": "spineGeneric"

            .. tab:: Command Line Interface

                Note that you can also pass this argument via CLI (see `Usage <../usage.html>`__)

                .. code-block:: bash

                    ivadomed -c path/to/config --path-output path/to/output/directory

    - ``loader_parameters:path_data``: Location of the dataset. As discussed in `Data <../data.html>`__, the dataset
      should conform to the BIDS standard. Modify the path so it points to the location of the downloaded dataset.

        .. tabs::

            .. tab:: JSON File

>                 At <`this line <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L12>`__ in the ``config.json`` is where you can update the ``path_data`` within the ``loader_parameters`` sub-dictionary.

Reviewer: ```suggestion
                At `this line <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L12>`__ in the ``config.json`` is where you can update the ``path_data`` within the ``loader_parameters`` sub-dictionary.
```

                .. code-block:: json

                    "slice_axis": "axial"

    - ``loader_parameters:multichannel``: Turn on/off multi-channel training. If ``true``, each sample has several
      channels, where each channel is an image contrast. If ``false``, only one image contrast is used per sample.

        .. tabs::

            .. tab:: JSON File

                At `this line <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L30>`__ in the ``config.json`` is where you can update the ``multichannel`` subkey within the ``loader_parameters`` sub-dictionary.

                .. code-block:: json

                    "multichannel": false

                .. note::

                    The multichannel approach requires that for each subject, the image contrasts are co-registered. This implies that
                    a ground truth segmentation is aligned with all contrasts, for a given subject. In this tutorial, only one channel
                    will be used.

>     - ``training_time:num_epochs``: the maximum number of epochs that will be run during training. Each epoch is composed

Reviewer: ```suggestion
    - ``training_parameters:training_time:num_epochs``: the maximum number of epochs that will be run during training. Each epoch is composed
```

      of a training part and an evaluation part. It should be a strictly positive integer.

        .. tabs::

            .. tab:: JSON File

>                 At `this line <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L29>`__ in the ``config.json`` is where you can update the ``slice_axis`` subkey within the ``loader_parameters`` sub-dictionary.

Reviewer: ```suggestion
                At `this line <https://github.com/ivadomed/ivadomed/blob/master/ivadomed/config/config.json#L48>`__ in the ``config.json`` is where you can update the ``num_epochs`` subkey within the ``training_parameters:training_time`` sub-dictionary.
```

Train model
-----------

    Once the configuration file is ready, run the training:

    .. tabs::

        .. tab:: Command Line Interface

            .. code-block:: bash

               ivadomed --train -c config.json --path-data path/to/bids/data --path-output path/to/output/directory

            - In the above command, we execute the ``--train`` command and manually specified ``--path-data`` and ``--path-output`` and overwrote/replace the specification in ``config.json``

            - ``--train``: We can pass other flags to execute different commands (training, testing, segmentation), see `Usage <../usage.html>`__.

            - ``--path-output``: Folder name that will contain the output files (e.g., trained model, predictions, results).

            - ``--path-data``: Location of the dataset. As discussed in `Data <../data.html>`__, the dataset
              should conform to the BIDS standard. Modify the path so it points to the location of the downloaded dataset.

        .. tab:: JSON File

            If you set the ``command``, ``path_output``, and ``path_data`` arguments in your config file, you do not need to pass the above the specific CLI flags.

>             Instead, make the following changes to the JSOn file at the specific lines:

Reviewer: ```suggestion
            Instead, make the following changes to the JSON file at the specific lines:
```

    To test the trained model on the testing sub-dataset and compute evaluation metrics, run:

    .. tabs::

        .. tab:: Command Line Interface

            .. code-block:: bash

               ivadomed --test -c config.json --path-data path/to/bids/data --path-output path/to/output/directory

        .. tab:: JSON File

            If you prefer to use config files over CLI flags, set ``command`` to the following in you config file:

            .. code-block:: json

               "command": "test"

>             You can also set ``path_output``, and ``path_data`` arguments in your config file, usually around **Line Number 2, 4, and 12** in the `config.json` respectively.

Reviewer: We shouldn't refer to specific line numbers here.
I don't think we need to add a link either, as those argument are widely discuss beforehand in the tutorial.

Author: Good point. I agree.

    To test the trained model on the testing sub-dataset and compute evaluation metrics, run:

    .. tabs::

        .. tab:: Command Line Interface

            .. code-block:: bash

               ivadomed --test -c config.json --path-data path/to/bids/data --path-output path/to/output/directory

        .. tab:: JSON File

            If you prefer to use config files over CLI flags, set ``command`` to the following in you config file:

            .. code-block:: json

               "command": "test"

            You can also set ``path_output``, and ``path_data`` arguments in your config file, usually around **Line Number 2, 4, and 12** in the `config.json` respectively.

    Then run:

    .. tabs::

        .. tab:: Command Line Interface

            .. code-block:: bash

>                 ivadomed -c config.json

Reviewer: the "Then run" part only applies if the "test" command is set in the JSON tab above, which is not clear here.

Author: Great point, it was an indentation issue, this should have been under the "JSON File" block and not gets elevated in level.
